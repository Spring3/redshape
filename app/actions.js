export const LOGIN = 'LOGIN';
const login = data => ({ type: LOGIN, data });

export const LOGOUT = 'LOGOUT';
const logout = () => ({ type: LOGOUT });

export const UPDATE_SETTINGS = 'UPDATE_SETTINGS';
const updateSettings = data => ({ type: UPDATE_SETTINGS, data });

export default {
  user: {
    login,
    logout
  },
  settings: {
    update: updateSettings
  }
};
