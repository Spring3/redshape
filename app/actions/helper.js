import * as axios from '../../modules/request';

function IssueFilter () {
  const filter = {};

  this.issue = (ids) => {
    filter.issue_id = Array.isArray(ids) ? ids.join(',') : ids;
    return this;
  };

  this.project = (id) => {
    filter.project_id = id;
    return this;
  };

  this.subproject = (id) => {
    filter.subprojectId = id;
    return this;
  };

  this.tracker = (id) => {
    filter.tracker_id = id;
    return this;
  };

  this.status = ({ open, closed }) => {
    if (open && closed) {
      filter.status_id = '*';
    } else {
      filter.status_id = open ? 'open' : 'closed';
    }
    return this;
  };

  this.assignee = (id) => {
    filter.assigned_to_id = id;
    return this;
  };

  this.author = (id) => {
    filter.author_id = id;
    return this;
  };

  this.title = (str) => {
    if (str) {
      filter.subject = encodeURIComponent(`~${str.trim()}`);
    }
    return this;
  };

  this.due = (date) => {
    filter.due_date = date.toISOString();
    return this;
  };

  this.createdAt = (date) => {
    filter.created_on = date.toISOString();
    return this;
  };

  this.createdBetween = (startDate, endDate) => {
    filter.created_on = encodeURIComponent(`><${startDate.toISOString()}|${endDate.toISOString()}`);
    return this;
  };

  this.priority = (id) => {
    filter.priority_id = id;
    return this;
  };

  this.sort = (by, direction) => {
    if (by && direction) {
      const column = by.indexOf('.') !== -1
        ? by.substring(0, by.indexOf('.'))
        : by;
      filter.sort = `${column}:${direction}`;
    }
    return this;
  };

  this.build = () => ({ ...filter });
}

const handleReject = (e) => {
  console.log('EEEE', e);
  // if this request was not cancelled
  if (!axios.default.isCancel(e)) {
    return Promise.reject(new Error(`Error ${e.status} (${e.statusText || e.message})`));
  }
  console.log('Request was cancelled');
  return undefined;
};

const request = ({
  url,
  method = 'GET',
  data,
  headers,
  query,
  id
}) => {
  const requestConfig = {
    url,
    method,
    headers: axios.makeCancellable(headers || {}, id)
  };

  if (!['GET', 'DELETE'].includes(method)) {
    requestConfig.data = typeof data === 'string' ? data : JSON.stringify(data);
  }

  if (query) {
    requestConfig.params = query;
  }

  if (!axios.getInstance()) {
    return Promise.reject(new Error('401 - Unauthorized'));
  }

  return axios.getInstance().request(requestConfig)
    .then((res) => {
      delete axios.pendingRequests[id];
      return ({ data: res.data });
    })
    .catch((error) => {
      delete axios.pendingRequests[id];
      return handleReject(error);
    });
};

const login = ({
  redmineEndpoint,
  url,
  headers
}) => {
  const defaultConfig = axios.getDefaultConfig();
  return axios.default.request({
    baseURL: redmineEndpoint,
    timeout: defaultConfig.timeout,
    headers: {
      ...defaultConfig.headers,
      ...(headers || {})
    },
    url,
    method: 'GET'
  }).then((res) => {
    axios.reset();
    const { api_key } = res.data.user || {};
    if (api_key) {
      axios.initialize(redmineEndpoint, api_key);
    }
    return { data: res.data };
  });
};

const notify = {
  start: (type, id) => ({ type, status: 'START', id }),
  paginate: (type, data, id) => ({ type, data, status: 'PAGE_NEXT', id }),
  ok: (type, data, id) => ({ type, data, status: 'OK', id }),
  nok: (type, data, id) => ({ type, data, status: 'NOK', id }),
  cancel: type => ({ type, status: 'CANCELLED' })
};

export {
  IssueFilter,
  notify,
  login
};

export default request;
