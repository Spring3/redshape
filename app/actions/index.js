import userActions from './user.actions';
import trackingActions from './tracking.actions';
import issuesActions from './issues.actions';
import projectActions from './project.actions';
import timeEntryActions from './timeEntry.actions';
import settingsActions from './settings.actions';

export default {
  user: userActions,
  issues: issuesActions,
  tracking: trackingActions,
  projects: projectActions,
  timeEntry: timeEntryActions,
  settings: settingsActions
};
