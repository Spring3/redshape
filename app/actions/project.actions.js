import request, { notify } from './helper';

export const PROJECT_GET_ALL = 'PROJECT_GET_ALL';

const getAll = () => async (dispatch) => {
  dispatch(notify.start(PROJECT_GET_ALL));

  const requestPage = (offset = 0) => request({
    url: '/projects.json',
    query: {
      include: 'time_entry_activities',
      offset
    },
    id: `getProjects:${offset}`
  }).then(({ data }) => data);

  return requestPage()
    .then((data) => {
      const { total_count, offset, limit } = data;
      if (total_count > (offset + limit)) {
        // Math.floor because we already fetched the 1st page.
        const remainingPages = Math.floor(total_count / limit);
        const promises = [data];
        for (let i = 0; i < remainingPages; i++) {
          promises.push(requestPage((i + 1) * limit));
        }
        return Promise.all(promises);
      }
      return [data];
    })
    .then((results) => {
      dispatch(notify.ok(PROJECT_GET_ALL, results));
    })
    .catch((error) => {
      console.error('Error when trying to get the info about the existing projects', error);
      dispatch(notify.nok(PROJECT_GET_ALL, error));
    });
};

export default {
  getAll
};
