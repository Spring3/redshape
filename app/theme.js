export default {
  bg: '#FFFFFF',
  bgLight: '#F9F9F9',
  mainLight: '#FF7F87',
  main: '#FF7079',
  mainDark: '#EF6069',
  minorText: '#909090',
  normalText: '#3F3844',
  hoverText: '#FFFFFF',
  red: '#FF634D',
  yellow: '#FFDA77',
  green: '#6CCA51',
  transitionTime: '.2s'
};
