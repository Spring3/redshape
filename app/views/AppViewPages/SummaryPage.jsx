import React, { Component } from 'react';
import PropTypes from 'prop-types';
import debounce from 'lodash/debounce';
import { connect } from 'react-redux';
import styled from 'styled-components';

import { IssueFilter } from '../../actions/helper';
import actions from '../../actions';
import { Input } from '../../components/Input';
import IssuesTable from '../../components/SummaryPage/IssuesTable';
import OptionsBlock from '../../components/SummaryPage/OptionsBlock';
import ColumnHeadersSelect from '../../components/SummaryPage/ColumnHeadersSelect';

const Grid = styled.div`
  display: grid;
  padding: 20px;
  grid-template-columns: repeat(auto-fit, minmax(100px, 1fr));
  grid-auto-rows: 100px;
  grid-auto-flow: dense;
  grid-gap: 20px;
  margin-bottom: 60px;
`;

const Section = styled.section`
  background: white;
  padding: 0px 20px 20px 20px;
  border-radius: 5px;
`;

const IssuesSection = styled(Section)`
  grid-column: span 8;
  grid-row: span 6;
`;

const ActivitySection = styled(Section)`
  grid-column: span 3;
  grid-row: span 4;
`;

const OptionsGrid = styled.div`
  display: grid;
  grid-template-columns: minmax(200px, auto) 1fr;
  grid-template-rows: auto;
  grid-row-gap: 20px;
  grid-column-gap: 20px;
  margin-bottom: 20px;
`;

const GridRow = styled.div`
  grid-column: 1/-1;
`;

class SummaryPage extends Component {
  constructor(props) {
    super(props);

    this.state = {
      search: undefined,
      sortBy: undefined,
      sortDirection: undefined
    };

    this.deboucedFetch = debounce(this.fetchIssues, 500);
  }

  componentWillMount() {
    this.fetchIssues();
  }

  componentDidUpdate(oldProps) {
    if (oldProps.showClosedIssues !== this.props.showClosedIssues) {
      this.fetchIssues();
    }
  }

  fetchIssues = (page = 0) => {
    const { search, sortBy, sortDirection } = this.state;
    const { userId, showClosedIssues } = this.props;
    const queryFilter = new IssueFilter()
      .assignee(userId)
      .status({ open: true, closed: showClosedIssues })
      .title(search)
      .sort(sortBy, sortDirection)
      .build();
    this.props.fetchIssues(queryFilter, page);
  }

  onSearchChange = (e) => {
    this.setState({
      search: e.target.value
    }, () => this.deboucedFetch());
  }

  onSort = (sortBy, sortDirection) => {
    this.setState({
      sortBy: sortBy,
      sortDirection: sortDirection
    }, () => this.deboucedFetch());
  }

  render() {
    return (
      <Grid>
        <IssuesSection>
          <h2>Issues assigned to me</h2>
          <OptionsGrid>
            <OptionsBlock />
            <ColumnHeadersSelect />
            <GridRow>
              <Input
                type="text"
                name="search"
                placeholder="Search"
                onChange={this.onSearchChange}
              />
            </GridRow>
          </OptionsGrid>
          <IssuesTable
            onSort={this.onSort}
            fetchIssuePage={this.fetchIssues}
          />
        </IssuesSection>
        <ActivitySection>
          <h2>Activity Stack</h2>
        </ActivitySection>
      </Grid>
    );
  }
}

SummaryPage.propTypes = {
  userId: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number
  ]).isRequired,
  showClosedIssues: PropTypes.bool.isRequired,
  fetchIssues: PropTypes.func.isRequired
};

const mapStateToProps = state => ({
  userId: state.user.id,
  showClosedIssues: state.settings.showClosedIssues
});

const mapDispatchToProps = dispatch => ({
  fetchIssues: (filter, page) => dispatch(actions.issues.getPage(filter, page))
});

export default connect(mapStateToProps, mapDispatchToProps)(SummaryPage);
