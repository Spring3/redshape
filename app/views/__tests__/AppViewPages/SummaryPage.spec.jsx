import React from 'react';
import { HashRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import MockAdapter from 'axios-mock-adapter';
import { ThemeProvider } from 'styled-components';

import { initialize, getInstance, reset } from '../../../../modules/request';
import theme from '../../../theme';
import SummaryPage from '../../AppViewPages/SummaryPage';
import { ISSUES_GET_ALL } from '../../../actions/issues.actions';

const mockStore = configureStore([thunk]);
const redmineEndpoint = 'redmine.test.test';
const token = 'multipass';

let axiosMock;

const fixtures = {
  issues: [
    {
      id: '1',
      project: { name: 'Test Project 1' },
      tracker: { name: 'Test Tracker 1' },
      status: { name: 'Open' },
      subject: 'Task #1',
      priority: { name: 'High' },
      estimated_hours: '10',
      due_date: new Date().toISOString()
    },
    {
      id: '2',
      project: { name: 'Test Project 2' },
      tracker: { name: 'Test Tracker 2' },
      status: { name: 'Open' },
      subject: 'Task #2',
      priority: { name: 'Normal' },
      estimated_hours: '5',
      due_date: new Date().toISOString()
    },
    {
      id: '3',
      project: { name: 'Test Project 3' },
      tracker: { name: 'Test Tracker 3' },
      status: { name: 'Open' },
      subject: 'Task #3',
      priority: { name: 'Low' },
      estimated_hours: '2',
      due_date: new Date().toISOString()
    }
  ], // Id and Subject are fixed, so they go first
  expectedHeaders: ['Id', 'Subject', 'Project', 'Tracker', 'Status', 'Priority', 'Estimation', 'Due Date'],
  valueMapping: {
    Id: 'id',
    Project: 'project.name',
    Tracker: 'tracker.name',
    Status: 'status.name',
    Subject: 'subject',
    Priority: 'priority.name',
    Estimation: 'estimated_hours',
    'Due Date': 'due_date'
  }
};

describe('AppView -> Summary Page', () => {
  beforeAll(() => {
    initialize(redmineEndpoint, token);
    axiosMock = new MockAdapter(getInstance());
  });

  afterEach(() => {
    axiosMock.reset();
  });

  afterAll(() => {
    axiosMock.restore();
    reset();
  });

  it('should match the snapshot', () => {
    axiosMock.onGet('/issues.json').reply(() => Promise.resolve([200, {}]));
    const store = mockStore({
      user: {
        id: 1,
        firstname: 'firstname',
        lastname: 'lastname',
        redmineEndpoint: 'https://redmine.domain',
        api_key: '123abc'
      },
      issues: {
        all: {
          data: []
        }
      },
      settings: {
        issueHeaders: [
          { label: 'Id', isFixed: true, value: 'id' },
          { label: 'Subject', isFixed: true, value: 'subject' }
        ]
      }
    });
    const tree = renderer.create(
      <Provider store={store}>
        <HashRouter>
          <ThemeProvider theme={theme}>
            <SummaryPage />
          </ThemeProvider>
        </HashRouter>
      </Provider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should fetch the issues on mount', () => {
    axiosMock.onGet('/issues.json').reply(() => Promise.resolve([200, {}]));
    const state = {
      user: {
        id: 1,
        firstname: 'firstname',
        lastname: 'lastname',
        redmineEndpoint: 'https://redmine.domain',
        api_key: '123abc'
      },
      issues: {
        all: {
          data: []
        }
      },
      settings: {
        issueHeaders: [
          { label: 'Id', isFixed: true, value: 'id' },
          { label: 'Subject', isFixed: true, value: 'subject' }
        ]
      }
    };
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <ThemeProvider theme={theme}>
          <HashRouter>
            <SummaryPage />  
          </HashRouter>
        </ThemeProvider>
      </Provider>
    );

    expect(store.getActions()[0].type).toBe(ISSUES_GET_ALL);
  });

  it('should fetch issues on search', (done) => {
    axiosMock.onGet('/issues.json').reply(() => Promise.resolve([200, {}]));
    const state = {
      user: {
        id: 1,
        firstname: 'firstname',
        lastname: 'lastname',
        redmineEndpoint: 'https://redmine.domain',
        api_key: '123abc'
      },
      issues: {
        all: {
          data: []
        }
      },
      settings: {
        issueHeaders: [
          { label: 'Id', isFixed: true, value: 'id' },
          { label: 'Subject', isFixed: true, value: 'subject' }
        ]
      }
    };
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <ThemeProvider theme={theme}>
          <HashRouter>
            <SummaryPage />
          </HashRouter>
        </ThemeProvider>
      </Provider>
    );

    const page = wrapper.find('SummaryPage').instance();
    expect(page).toBeTruthy();

    page.onSearchChange({ target: { value: 'search' } });
    expect(page.state.search).toBe('search');
    setTimeout(() => {
      expect(store.getActions().length).toBe(2);
      expect(store.getActions().filter(action => action.type === ISSUES_GET_ALL).length).toBe(2);
      done();
    }, 1);
  });

  it('should set the sorting properties', (done) => {
    axiosMock.onGet('/issues.json').reply(() => Promise.resolve([200, {}]));
    const state = {
      user: {
        id: 1,
        firstname: 'firstname',
        lastname: 'lastname',
        redmineEndpoint: 'https://redmine.domain',
        api_key: '123abc'
      },
      issues: {
        all: {
          data: []
        }
      },
      settings: {
        issueHeaders: [
          { label: 'Id', isFixed: true, value: 'id' },
          { label: 'Subject', isFixed: true, value: 'subject' }
        ]
      }
    };
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <ThemeProvider theme={theme}>
          <HashRouter>
            <SummaryPage />
          </HashRouter>
        </ThemeProvider>
      </Provider>
    );

    const page = wrapper.find('SummaryPage').instance();
    expect(page).toBeTruthy();

    page.onSort('id', 'desc');
    setTimeout(() => {
      expect(page.state.sortBy).toBe('id');
      expect(page.state.sortDirection).toBe('desc');
      expect(store.getActions().length).toBe(2);
      expect(store.getActions().filter(action => action.type === ISSUES_GET_ALL).length).toBe(2);
      done();
    }, 1);
  });

  it('should fetch the issues when showClosedIssues state changes', (done) => {
    axiosMock.onGet('/issues.json').reply(() => Promise.resolve([200, {}]));
    const state = {
      user: {
        id: 1,
        firstname: 'firstname',
        lastname: 'lastname',
        redmineEndpoint: 'https://redmine.domain',
        api_key: '123abc'
      },
      issues: {
        all: {
          data: []
        }
      },
      settings: {
        issueHeaders: [
          { label: 'Id', isFixed: true, value: 'id' },
          { label: 'Subject', isFixed: true, value: 'subject' }
        ]
      }
    };
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <ThemeProvider theme={theme}>
          <HashRouter>
            <SummaryPage />
          </HashRouter>
        </ThemeProvider>
      </Provider>
    );

    const page = wrapper.find('SummaryPage').instance();
    expect(page).toBeTruthy();

    page.setState({ showClosedIssues: true });
    setTimeout(() => {
      expect(page.state.showClosedIssues).toBe(true);
      expect(store.getActions().length).toBe(2);
      expect(store.getActions().filter(action => action.type === ISSUES_GET_ALL).length).toBe(2);
      done();
    }, 1);
  });
    // it('render issues in the table', () => {
    //   const { issues, expectedHeaders, valueMapping } = fixtures;
    //   axiosMock.onGet('/issues.json').reply(() => Promise.resolve([200, { issues }]));
    //   const store = mockStore({
    //     user: {
    //       id: 1,
    //       firstname: 'firstname',
    //       lastname: 'lastname',
    //       redmineEndpoint: 'https://redmine.domain',
    //       api_key: '123abc'
    //     },
    //     issues: {
    //       all: {
    //         data: issues
    //       }
    //     },
    //     settings: {
    //       issueHeaders: [
    //         { label: 'Id', isFixed: true, value: 'id' },
    //         { label: 'Subject', isFixed: true, value: 'subject' }
    //       ]
    //     }
    //   });
    //   const wrapper = mount(
    //     <Provider store={store}>
    //       <HashRouter>
    //         <ThemeProvider theme={theme}>
    //           <SummaryPage />
    //         </ThemeProvider>
    //       </HashRouter>
    //     </Provider>
    //   );

    //   const summaryPage = wrapper.find('SummaryPage');
    //   expect(summaryPage.state('issues')).toEqual(issues);

    //   const tableRows = wrapper.find('tr');
    //   expect(tableRows.length).toBe(issues.length + 1); // because table headers take 1 tr

    //   const tableHeaders = tableRows.at(0).children();

    //   for (let i = 0; i < expectedHeaders.length; i++) {
    //     expect(tableHeaders.at(i).text()).toBe(expectedHeaders[i]);
    //   }

    //   // -1 to not include the title row
    //   for (let i = 0; i < tableRows.length - 1; i++) {
    //     const tableData = tableRows.at(i + 1).children();
    //     for (let j = 0; j < tableData.length; j++) {
    //       expect(tableData.at(j).text()).toBe(_.get(issues[i], valueMapping[expectedHeaders[j]]));
    //     }
    //   }
    // });

    // it('should be able to search the items in the issues table', () => {
    //   const { issues, expectedHeaders, valueMapping } = fixtures;
    //   fetch.mockResponse(JSON.stringify({ issues }));
    //   const store = mockStore({
    //     user: {
    //       id: 1,
    //       firstname: 'firstname',
    //       lastname: 'lastname',
    //       redmineEndpoint: 'https://redmine.domain',
    //       api_key: '123abc'
    //     },
    //     issues: {
    //       all: {
    //         data: issues
    //       }
    //     }
    //   });
    //   const wrapper = mount(
    //     <Provider store={store}>
    //       <HashRouter><SummaryPage /></HashRouter>
    //     </Provider>
    //   );

    //   let summaryPage = wrapper.find('SummaryPage');
    //   expect(summaryPage.state('issues')).toEqual(issues);
    //   const input = wrapper.find('input[name="search"]');
    //   expect(input.exists()).toBeTruthy();

    //   input.simulate('change', { target: { value: issues[1].subject } });
    //   wrapper.update();

    //   summaryPage = wrapper.find('SummaryPage');
    //   expect(summaryPage.state('issues')).toEqual(issues.slice(1, 2));

    //   const tableRows = wrapper.find('tr');
    //   expect(tableRows.length).toBe(2); // because table headers take 1 tr + 1 row with result

    //   // starting at 1 to not include the title row
    //   const tableData = tableRows.at(1).children();
    //   for (let j = 0; j < tableData.length; j++) {
    //     expect(tableData.at(j).text()).toBe(_.get(issues[1], valueMapping[expectedHeaders[j]]));
    //   }
    // });

    // it('should be able to include closed tasks in the table', () => {
    //   const { issues } = fixtures;
    //   fetch.mockResponse(JSON.stringify({ issues }));
    //   const store = mockStore({
    //     user: {
    //       id: 1,
    //       firstname: 'firstname',
    //       lastname: 'lastname',
    //       redmineEndpoint: 'https://redmine.domain',
    //       api_key: '123abc'
    //     },
    //     issues: {
    //       all: {
    //         data: issues
    //       }
    //     }
    //   });
    //   const wrapper = mount(
    //     <Provider store={store}>
    //       <HashRouter><SummaryPage /></HashRouter>
    //     </Provider>
    //   );

    //   const summaryPage = wrapper.find('SummaryPage');
    //   expect(summaryPage.state('issues')).toEqual(issues);
    //   expect(summaryPage.state('showClosed')).toBe(false);

    //   const checkbox = wrapper.find('input[name="showClosed"]');
    //   expect(checkbox.exists()).toBeTruthy();

    //   checkbox.simulate('change', { target: { checked: true } });
    //   wrapper.update();

    //   expect(summaryPage.state('showClosed')).toBe(true);
    //   expect(store.getActions().length).toBe(2);
    // });
});
