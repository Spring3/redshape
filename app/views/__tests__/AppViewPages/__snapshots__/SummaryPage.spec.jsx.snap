// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppView -> Summary Page should match the snapshot 1`] = `
.c11 {
  display: block;
  width: 100%;
  border-radius: 3px;
  padding: 5px 10px;
  box-sizing: border-box;
  font-size: 14px;
  min-height: 35px;
  outline: none;
  font-weight: bold;
  -webkit-transition: background .2s;
  transition: background .2s;
  border: 1px solid #909090;
  color: #FF7079;
  background: white;
}

.c11:hover {
  border: 1px solid #FF7079;
}

.c11:focus {
  border-color: #FF7079;
  box-shadow: 0px 0px 0px 1px #FF7079;
}

.c11::-webkit-input-placeholder {
  color: #909090;
}

.c11::-moz-placeholder {
  color: #909090;
}

.c11:-ms-input-placeholder {
  color: #909090;
}

.c11::placeholder {
  color: #909090;
}

.c7 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  -webkit-clippath: inset(50%);
  clippath: inset(50%);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

.c8 {
  display: inline-block;
  width: 16px;
  height: 16px;
  background: #F9F9F9 transition:background .2s;
  border: 2px solid transparent;
  border-radius: 3px;
  cursor: pointer;
  background: white;
  border-color: #FF7079;
}

.c8 svg {
  visibility: hidden;
  position: relative;
  vertical-align: middle;
  bottom: 3px;
}

.c6:not(:disabled):hover + .c8,
.c6:focus + .c8 {
  box-shadow: 0px 0px 5px 1px #FF7F87;
}

.c6:disabled + .c8 {
  background: #F9F9F9;
  border-color: lightgrey;
}

.c5 {
  display: inline-block;
  vertical-align: middle;
}

.c5 + span {
  margin-left: 10px;
  vertical-align: middle;
}

.c3 h4 {
  margin-bottom: 10px;
  color: #909090;
}

.c4 {
  color: #909090;
}

.c12 {
  position: relative;
}

.c12 tbody {
  text-align: center;
  font-size: 14px;
  overflow-y: scroll;
  display: block;
  max-height: 60vh;
}

.c12 tbody tr th {
  padding: 15px 5px;
  position: -webkit-sticky;
  position: sticky;
  top: 0;
  background: #F9F9F9;
  font-weight: 900;
  -webkit-transition: color ease .2s;
  transition: color ease .2s;
}

.c12 tbody tr th:hover {
  cursor: pointer;
  color: #FF7079;
}

.c12 tbody tr th svg {
  vertical-align: middle;
}

.c12 tbody tr th.due-date {
  width: 90px;
  z-index: 1;
}

.c12 tbody tr td {
  padding: 15px 5px;
  font-weight: 500;
  -webkit-transition: background ease .2s;
  transition: background ease .2s;
}

.c12 tbody tr:hover {
  cursor: pointer;
}

.c12 tbody tr:hover td {
  background: #F9F9F9;
}

.c9 {
  margin-top: 10px;
}

.c0 {
  display: grid;
  padding: 20px;
  grid-template-columns: repeat(auto-fit,minmax(100px,1fr));
  margin-bottom: 60px;
}

.c1 {
  background: white;
  padding: 0px 20px 20px 20px;
  border-radius: 5px;
  grid-column: span 8;
  grid-row: auto;
}

.c2 {
  display: grid;
  grid-template-columns: minmax(200px,auto) 1fr;
  grid-template-rows: auto;
  grid-row-gap: 20px;
  grid-column-gap: 20px;
  margin-bottom: 20px;
}

.c10 {
  grid-column: 1/-1;
}

<div
  className="c0"
>
  <section
    className="c1"
  >
    <h2>
      Issues assigned to me
    </h2>
    <div
      className="c2"
    >
      <div>
        <div
          className="form-group undefined c3"
        >
          <h4
            className="c4"
            htmlFor="queryOptions"
          >
            Options
          </h4>
          <div
            id="queryOptions"
          >
            <label>
              <div
                className="c5"
              >
                <input
                  checked={false}
                  className="c6 c7"
                  disabled={false}
                  onChange={[Function]}
                  type="checkbox"
                />
                <div
                  checked={false}
                  className="c8"
                >
                  <svg
                    className="mdi-icon "
                    fill="white"
                    height="18"
                    viewBox="0 0 24 24"
                    width="18"
                  >
                    <path
                      d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"
                    />
                  </svg>
                </div>
              </div>
              <span>
                Include Closed
              </span>
            </label>
          </div>
        </div>
        <div
          className="c9"
        >
          <label>
            <div
              className="c5"
            >
              <input
                checked={false}
                className="c6 c7"
                disabled={false}
                onChange={[Function]}
                type="checkbox"
              />
              <div
                checked={false}
                className="c8"
              >
                <svg
                  className="mdi-icon "
                  fill="white"
                  height="18"
                  viewBox="0 0 24 24"
                  width="18"
                >
                  <path
                    d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"
                  />
                </svg>
              </div>
            </div>
            <span>
              Use Colors
            </span>
          </label>
        </div>
      </div>
      <div
        className="form-group undefined c3"
      >
        <h4
          className="c4"
          htmlFor="headers"
        >
          Table Columns
        </h4>
        <div
          className="css-10nd86i"
          onKeyDown={[Function]}
        >
          <div
            className="css-1rcfx4g"
            onMouseDown={[Function]}
            onTouchEnd={[Function]}
          >
            <div
              className="css-1hwfws3"
            >
              <div
                style={
                  Object {
                    "overflow": "hidden",
                    "whiteSpace": "nowrap",
                    "width": "auto",
                  }
                }
              >
                <div
                  className="css-iuzuh3"
                >
                  <div
                    className="css-7ydjiw"
                  >
                    Id
                  </div>
                  <div
                    className="css-azd3r6"
                    onClick={[Function]}
                    onMouseDown={[Function]}
                    onTouchEnd={[Function]}
                  >
                    <svg
                      aria-hidden="true"
                      className="css-19bqh2r"
                      focusable="false"
                      height={14}
                      viewBox="0 0 20 20"
                      width={14}
                    >
                      <path
                        d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
              <div
                style={
                  Object {
                    "overflow": "hidden",
                    "whiteSpace": "nowrap",
                    "width": "auto",
                  }
                }
              >
                <div
                  className="css-iuzuh3"
                >
                  <div
                    className="css-7ydjiw"
                  >
                    Subject
                  </div>
                  <div
                    className="css-azd3r6"
                    onClick={[Function]}
                    onMouseDown={[Function]}
                    onTouchEnd={[Function]}
                  >
                    <svg
                      aria-hidden="true"
                      className="css-19bqh2r"
                      focusable="false"
                      height={14}
                      viewBox="0 0 20 20"
                      width={14}
                    >
                      <path
                        d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
              <div
                className="css-1g6gooi"
              >
                <div
                  className=""
                  style={
                    Object {
                      "display": "inline-block",
                    }
                  }
                >
                  <input
                    aria-autocomplete="list"
                    autoCapitalize="none"
                    autoComplete="off"
                    autoCorrect="off"
                    disabled={false}
                    id="react-select-2-input"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    spellCheck="false"
                    style={
                      Object {
                        "background": 0,
                        "border": 0,
                        "boxSizing": "content-box",
                        "color": "inherit",
                        "fontSize": "inherit",
                        "opacity": 1,
                        "outline": 0,
                        "padding": 0,
                        "width": "1px",
                      }
                    }
                    tabIndex="0"
                    type="text"
                    value=""
                  />
                  <div
                    style={
                      Object {
                        "height": 0,
                        "left": 0,
                        "overflow": "scroll",
                        "position": "absolute",
                        "top": 0,
                        "visibility": "hidden",
                        "whiteSpace": "pre",
                      }
                    }
                  >
                    
                  </div>
                </div>
              </div>
            </div>
            <div
              className="css-1wy0on6"
            >
              <span
                className="css-d8oujb"
              />
              <div
                aria-hidden="true"
                className="css-1ep9fjw"
                onMouseDown={[Function]}
                onTouchEnd={[Function]}
              >
                <svg
                  aria-hidden="true"
                  className="css-19bqh2r"
                  focusable="false"
                  height={20}
                  viewBox="0 0 20 20"
                  width={20}
                >
                  <path
                    d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                  />
                </svg>
              </div>
            </div>
          </div>
          <div>
            <input
              name="headers"
              type="hidden"
              value="id"
            />
            <input
              name="headers"
              type="hidden"
              value="subject"
            />
          </div>
        </div>
      </div>
      <div
        className="c10"
      >
        <input
          className="c11"
          disabled={false}
          name="search"
          onChange={[Function]}
          placeholder="Search"
          type="text"
        />
      </div>
    </div>
    <table
      cellPadding={0}
      cellSpacing={0}
      className="c12"
      width="100%"
    >
      <tbody>
        <tr>
          <th
            className=""
            onClick={[Function]}
          >
            Id
             
          </th>
          <th
            className=""
            onClick={[Function]}
          >
            Subject
             
          </th>
        </tr>
      </tbody>
    </table>
  </section>
</div>
`;
